File: src/layout/container/freeformcontainer/FreeFormContainer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package layout.container.freeformcontainer;

import java.util.HashMap;
import java.util.Map;

import javafx.scene.Node;

import layout.area.FreeFormArea;
import layout.Component;
import layout.container.Container;

/**
 */
public class FreeFormContainer extends Container<FreeFormArea> {

    private final Map<Integer,Component> children = new HashMap();


    @Override
    public Map<Integer, Component> getChildren() {
        return children;
    }

    @Override
    public void addChild(Integer index, Component c) {
        if(index == null) return;

        if(c==null) children.remove(index);
        else children.put(index, c);

        if(ui!=null) ui.loadWindow(index, c);
        setParentRec();
    }

    @Override
    public void removeChild(Integer index) {
        ui.closeWindow(index);
        children.remove(index);
    }


    @Override
    public Integer getEmptySpot() {
        return null;
    }

    @Override
    public Node load() {
        if(ui==null) ui = new FreeFormArea(this);
        ui.load();
        return ui.getRoot();
    }

    @Override
    public void show() {
//        super.show();
        if(ui!=null) ui.show();
    }

    @Override
    public void hide() {
//        super.hide();
//        if(gui !=null) gui.widgets.values().forEach(Area::hide);
                if(ui!=null) ui.hide();
    }

}

File: src/player/main/sp/it/pl/layout/container/FreeFormContainer.kt
package sp.it.pl.layout.container

import javafx.scene.Node
import sp.it.pl.layout.Component
import sp.it.pl.layout.FreeFormContainerDb
import sp.it.util.access.v
import sp.it.util.collections.filterNotNullValues
import java.util.HashMap
import sp.it.util.conf.cv
import sp.it.util.conf.def

class FreeFormContainer: Container<FreeFormContainerUi> {

   /** Name of this container. */
   override val name = "FreeFormContainer"
   /** Whether this container shows child window headers. Default true. */
   val showHeaders by cv(true).def(name = "Show header", info = "Whether window headers are visible")
   /** Children of this container by index */
   private val children = HashMap<Int, Component>()

   @JvmOverloads
   constructor(state: FreeFormContainerDb = FreeFormContainerDb()): super(state) {
      showHeaders.value = state.showHeaders
      children += state.children.mapValues { it.value?.toDomain() }.filterNotNullValues()
      setChildrenParents()
   }

   override fun getChildren(): Map<Int, Component> = children

   override fun addChild(index: Int?, c: Component?) {
      if (index==null) return

      if (c==null) children.remove(index)
      else children[index] = c

      ui?.load(index, c)
      setParentRec()
   }

   override fun removeChild(index: Int?) {
      if (index==null) return

      ui.closeWindow(index)
      getChildren()[index]?.close()
      children.remove(index)
      closeWindowIfEmpty()
   }

   override fun getEmptySpot() = null

   override fun load(): Node {
      if (ui==null) ui = FreeFormContainerUi(this)
      ui.load()
      return ui.root
   }

   override fun toDb() = FreeFormContainerDb(id, loadType.value, locked.value, showHeaders.value, children.mapValues { it.value.toDb() }, properties)

   override fun show() = ui?.show() ?: Unit

   override fun hide() = ui?.hide() ?: Unit

}